### ISSUE TITLE
[BUG] Objects.prototype.valueOf object is not loosely equal to similar values

### ISSUE THREAD

**Issue by ronald-mz** [tapjs/tapjs#971]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Have you read the `CONTRIBUTING` guide on posting bugs, and `CODE_OF_CONDUCT`?

- [X] yes I read the things

### This issue exists in the latest tap version

- [X] I am using the latest tap

### Description

Object instances that overwrite `.valueOf` to a primitive value are not loosely compared. Native object wrappers for primitives work fine, but custom objects do not.

##### Passes
```javascript
test.same(1, new Number(1)) 
```
##### Fails
```javascript
class Numberish { typeOf() { return 1 } }
test.same(1, new Numberish()) 
```

### Reproduction

https://github.com/tapjs/tapjs/pull/970

### Environment

It's a package install rather than global install

```console
$ npm ls tap
revtts@0.0.1 /home/rmz/.mnt/c7eb854e-0301-475c-997d-b85969ddca0e/Git/revtts
├─┬ @revtts/cedalion@npm:cedalion@0.0.1 -> ./lib/cedalion
│ └── tap@18.5.7

$ uname -a
Linux z 6.5.6-76060506-generic #202310061235~1697396945~22.04~9283e32 SMP PREEMPT_DYNAMIC Sun O x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by isaacs** [comment]
So to be clear, the suggestion is that they _should_ be considered loosely equal?

seems reasonable to me. Take a look at the `compare` methods on the classes in src/tcompare/src/*.ts in this repo. Should be fairly straightforward to check .valueOf at the appropriate places.

**Comment by ronald-mz** [comment]
@isaacs Hah. I got through the whole explanation and never said that clearly.

Yes. My suggestion is that it *should* be loosely equal.

And, do you mean that I should look at the `test` methods in those classes?


Also, I'm going to show a bit of ignorance. Is `Same` the right method for loose equivalence? I'm looking at the test method for `Same` and am not sure why some bits of logic are the way they are. The method *appears* to be something other than loose equivalence

For instance, when it's testing for symbols, the logic looks to be something akin to:

```javascript
const symbolA = Symbol('Foo')
const symbolB = Symbol('Foo')

return typeof a === typeof b && a.toString() === b.toString() // Returns true
```

But that's not what a loose equivalence of a Symbol would result in.

```javascript
const symbolA = Symbol('Foo')
const symbolB = Symbol('Foo')

return symbolA == symbolB // returns false
```

This is because a [new symbol is created every time for none registered symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol#description). And there are other bits of logic that don't line up with what I'd usually think is a loose equivalency.

**Comment by isaacs** [comment]
`Same` is sort of slightly looser than `==`. It's a deep equivalence, so it uses == for scalar types, doesn't require constructors to match, etc. Symbols are compared using their string values, and objects have to have all the same properties. `Has` is like `Same`, but the tested object just has to have all the matching props from the pattern, and can have other props as well. `Match` is total loosey goosey, `1` matches `Number` etc. Useful when you want to assert a bunch of things without knowing the specific values. 

Testing against valueOf in Same and Match makes sense to me. Not in SameStrict tho.

**Comment by isaacs** [comment]
> And, do you mean that I should look at the test methods in those classes?

Oh, yeah, maybe that's it. Sorry, on a phone keyboard. Whatever method it is with that giant ternary list of conditions.

**Comment by ronald-mz** [comment]
> Oh, yeah, maybe that's it. Sorry, on a phone keyboard. Whatever method it is with that giant ternary list of conditions.

Yup! That's test! Seems, I'm in the right place then.



> `Same` is sort of slightly looser than `==`. It's a deep equivalence, so it uses == for scalar types, doesn't require constructors to match, etc. Symbols are compared using their string values, and objects have to have all the same properties. `Has` is like `Same`, but the tested object just has to have all the matching props from the pattern, and can have other props as well. `Match` is total loosey goosey, `1` matches `Number` etc. Useful when you want to assert a bunch of things without knowing the specific values.
> 
> Testing against valueOf in Same and Match makes sense to me. Not in SameStrict tho.

Thank you for the clarification. Was confused for a bit there.

Also, I'll not get the commit out tonight, gonna see about getting it up sometime tomorrow though.

**Comment by ronald-mz** [comment]
Alright, seems I'm a liar. Made the change tonight.

Just before closing the computer I realized the logic would be far easier if I used optional chaining.

**Comment by isaacs** [comment]
Closed by #970

**Comment by ronald-mz** [comment]
Thank you for working with me on this issue and for providing such a useful tool!
